# End-to-End Machine Learning Exercise
## Predicting the fine amount of a parking citation

This exercise utilizes the LA Parking Citation data, maintained by Kaggle.

### Problem statement: Use the provided Los Angeles Parking Citation data to develop a model that can predict the fine amount, and then interpret the results to find the variables that are most predictive of the score.

This exercise is a supervised machine learning task, in that we are provided a set of data with target values - the "Fine amount" - and we want to train a model that will map a set of features to the target.

Since the outcome/target is a continuous variable (as opposed to categorical), predicting the "Fine amount" amounts to a regression problem - a class of machine learning problems that yields a continuous outcome.  

First, let's import some dependencies:

In [3]:
import sys
# !{sys.executable} -m pip install pyproj
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=RuntimeWarning)

In [4]:
import os
import math
import pandas as pd
import numpy as np
from bokeh.io import output_notebook, show
from bokeh.plotting import figure, gmap
from bokeh.models import ColumnDataSource, GMapOptions, CategoricalColorMapper, HoverTool
from bokeh.palettes import Spectral7
from bokeh.core.properties import value
import pyproj
from datetime import date
import calendar
from sklearn.model_selection import train_test_split, RandomizedSearchCV, cross_val_score
from sklearn.linear_model import LinearRegression, Lasso, Ridge, BayesianRidge
from sklearn.metrics import mean_squared_error
from sklearn.feature_selection import RFECV
from sklearn.preprocessing import StandardScaler
from sklearn.svm import LinearSVR, SVR
from sklearn.ensemble import RandomForestRegressor
import xgboost as xgb
from time import time

# 1. Read and Examine Data

In [5]:
pc_raw = pd.read_csv("parking-citations.csv")
pc_raw.head()

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (0,7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Ticket number           Issue Date  Issue time Meter Id  Marked Time  \
0    1103341116  2015-12-21T00:00:00      1251.0      NaN          NaN   
1    1103700150  2015-12-21T00:00:00      1435.0      NaN          NaN   
2    1104803000  2015-12-21T00:00:00      2055.0      NaN          NaN   
3    1104820732  2015-12-26T00:00:00      1515.0      NaN          NaN   
4    1105461453  2015-09-15T00:00:00       115.0      NaN          NaN   

  RP State Plate  Plate Expiry Date  VIN  Make Body Style Color  \
0             CA           200304.0  NaN  HOND         PA    GY   
1             CA           201512.0  NaN   GMC         VN    WH   
2             CA           201503.0  NaN  NISS         PA    BK   
3             CA                NaN  NaN  ACUR         PA    WH   
4             CA           200316.0  NaN  CHEV         PA    BK   

             Location  Route  Agency Violation code Violation Description  \
0     13147 WELBY WAY  01521     1.0         4000A1    NO EVIDENCE OF REG   
1       525 S MAIN ST   1C51     1.0         4000A1    NO EVIDENCE OF REG   
2       200 WORLD WAY    2R2     2.0           8939            WHITE CURB   
3       100 WORLD WAY   2F11     2.0            000                17104h   
4  GEORGIA ST/OLYMPIC  1FB70     1.0          8069A  NO STOPPING/STANDING   

   Fine amount   Latitude  Longitude  
0         50.0    99999.0    99999.0  
1         50.0    99999.0    99999.0  
2         58.0  6439997.9  1802686.4  
3          NaN  6440041.1  1802686.2  
4         93.0    99999.0    99999.0

In [6]:
# A function to summarize missing values in the raw data
def missing_values_table(df):
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        return mis_val_table_ren_columns

print(missing_values_table(pc_raw))

Your selected dataframe has 19 columns.
There are 17 columns that have missing values.
                       Missing Values  % of Total Values
VIN                           8676980               99.8
Marked Time                   8403537               96.7
Meter Id                      6430920               74.0
Plate Expiry Date              791689                9.1
Route                           65261                0.8
Body Style                       8847                0.1
Make                             8722                0.1
Fine amount                      6488                0.1
Color                            4099                0.0
Issue time                       2572                0.0
Violation Description             867                0.0
Location                          850                0.0
RP State Plate                    765                0.0
Agency                            543                0.0
Issue Date                        534                0.0
L

#### There are a number of issues and observations to be address here:
1. "VIN", "Marked Time", and "Meter Id" are, for the most part, missing data. It would be prudent to drop these columns.
2. "Plate Expiry Date" is missing only about 9% of its data, but it doesn't make much sense to attempt to impute these missing data, so I'll drop this column too.
3. The variable "Issue time" appears to be of "float" type, and ought to be converted into an HH:MM format
4. The "Latitude" and "Longitude" variables are in their absolute forms and need to be converted for projection purposes. Moreover, a Lat/Long of (99999.0, 99999.0) does not exist, and ought to be treated as missing information.
5. It is reasonable to assume that "Violation Description" and "Violation code" are collinear, so I'll arbitrarily choose to drop the "Violation code" column.
6. It is reasonable to assume that "Location" is collinear with the "Latitude"/"Longitude" variables, so I'll drop that column too.
7. It appears that the "Ticket Number" column is merely an index for the row-wise data. It doesn't appear to serve any other purpose.
8. Some columns have mixed data types, so I'll explicitly specify all data types when reading in the pared-down data frame to avoid any confusion.

## 1.1 Read and examine data after eliminating problematic columns

In [7]:
pc_raw = None

dtypes = {
 'Issue Date': 'object',
 'Issue time': 'float64',
 'RP State Plate': 'object',
 'Make': 'object',
 'Body Style': 'object',
 'Color': 'object',
 'Route': 'object',
 'Agency': 'object',
 'Violation Description': 'object',
 'Fine amount': 'float64',
 'Latitude': 'float64',
 'Longitude': 'float64'}

cols2use = list(dtypes.keys())

pc = pd.read_csv("parking-citations.csv", usecols = cols2use, dtype = dtypes, parse_dates = ["Issue Date"])

pc.head()

Issue Date  Issue time RP State Plate  Make Body Style Color  Route Agency  \
0 2015-12-21      1251.0             CA  HOND         PA    GY  01521      1   
1 2015-12-21      1435.0             CA   GMC         VN    WH   1C51      1   
2 2015-12-21      2055.0             CA  NISS         PA    BK    2R2      2   
3 2015-12-26      1515.0             CA  ACUR         PA    WH   2F11      2   
4 2015-09-15       115.0             CA  CHEV         PA    BK  1FB70      1   

  Violation Description  Fine amount   Latitude  Longitude  
0    NO EVIDENCE OF REG         50.0    99999.0    99999.0  
1    NO EVIDENCE OF REG         50.0    99999.0    99999.0  
2            WHITE CURB         58.0  6439997.9  1802686.4  
3                17104h          NaN  6440041.1  1802686.2  
4  NO STOPPING/STANDING         93.0    99999.0    99999.0

# 2. Data Cleaning and Preparation

## 2.1 Missing values
Where "Issue time" and "Fine amount" are missing information, I will impute their respective median values.

In [8]:
# Fill missing "Issue time" values with median
pc["Issue time"] = pc["Issue time"].fillna(pc["Issue time"].median())

# Fill missing "Fine amount" values with median
pc["Fine amount"] = pc["Fine amount"].fillna(pc["Fine amount"].median())

## 2.2 Converting "Issue time" to HH:MM format

In [9]:
# Convert "Issue time" to HH:MM format
def convert_time(float_time):
    string_time = str(float_time)
    string_time = string_time.split('.')[0]
    string_time = string_time.zfill(4)
    string_time = string_time[:2] + ":" + string_time[2:4]
    return string_time

pc["Issue time"] = pc["Issue time"].apply(convert_time)

## 2.3 Clean/Format "Latitude" and "Longitude"
The latitude/longitude data are given in their absolute forms, and need to be converted in order to project onto x-y axes. The pyproj package will be very helpful in achieving this.

In [10]:
# First, convert the unknown (99999.000, 999999.000) coordinates to NaN and remove them
pc['Latitude'] = np.where(pc['Latitude']==99999.000, np.nan, pc['Latitude'])
pc['Longitude'] = np.where(pc['Longitude']==99999.000, np.nan, pc['Longitude'])
pc = pc.dropna(subset = ["Latitude"])

# Next, convert the remaining pairs of coordinates for x-y projection
pm = '+proj=lcc +lat_1=34.03333333333333 +lat_2=35.46666666666667 +lat_0=33.5 +lon_0=-118 +x_0=2000000 +y_0=500000.0000000002 +ellps=GRS80 +datum=NAD83 +to_meter=0.3048006096012192 +no_defs'
x1m,y1m = pc['Latitude'].values, pc['Longitude'].values
x2m,y2m = pyproj.transform(pyproj.Proj(pm,preserve_units = True), pyproj.Proj("+init=epsg:4326"), x1m,y1m)
pc['Latitude']=x2m
pc['Longitude']=y2m

In [11]:
# Move "Fine amount" to the first column of the data set, for easy feature/target selection in future steps
col_names = pc.columns.tolist()
col_names.remove("Fine amount")
col_names.insert(0,"Fine amount")
pc = pc[col_names]

In [12]:
pc.head()

Fine amount Issue Date Issue time RP State Plate  Make Body Style Color  \
2          58.0 2015-12-21      20:55             CA  NISS         PA    BK   
3          68.0 2015-12-26      15:15             CA  ACUR         PA    WH   
33         93.0 2015-12-21      10:41             CA  HOND         PA    BK   
42        363.0 2015-12-19      15:50             CA  HOND        NaN    GO   
43         68.0 2015-12-26      18:15             NY  TOYO         PA    WH   

    Route Agency Violation Description    Latitude  Longitude  
2     2R2      2            WHITE CURB -118.401151  33.945212  
3    2F11      2                17104h -118.401009  33.945212  
33  15EL5      1              RED ZONE -118.363082  34.185788  
42  35MQ1      1     HANDICAP/NO DP ID -118.438092  34.157607  
43    NaN     36       PK IN PROH AREA -118.311225  34.126932

#### At this point, we've dealt with most of the practically important missing data...
Moreover, imputing missing data for things like "Route" and "Make" doesn't make much sense, practically speaking.

#### Onwards, to data exploration and feature engineering!

# 3. Feature Engineering

## 3.1 "Agency"
The "Agency" variable contains mixed categories that appear to be inappropriately differentiated. For example, we'll assume that the agency labeled "56" is identical to the agency labeled "56.0". Let's fix this ...

In [13]:
# Re-engineer "Agency" as floats, then back to strings
pc["Agency"] = pc["Agency"].astype("float")
pc["Agency"] = pc["Agency"].astype("object")

## 3.2 Date/Time information extraction
Let's extract the hour of the day, the day of the week, the day number of the month, the month of the year, and the year all as distinct features.

In [14]:
pc["Hour"] = pd.to_datetime(pc["Issue time"]).dt.hour
pc["Year"] = pc["Issue Date"].dt.year
pc["Month"] = pc["Issue Date"].dt.month
pc["Day"] = pc["Issue Date"].dt.day

def get_dayofweek(datetime):
    return calendar.day_name[datetime.weekday()]

pc["DayOfWeek"] = pc["Issue Date"].apply(get_dayofweek)

## 3.3 Locale
We have valuable information on the offending party's license plate state, but let's see if a simple binary variable can capture the value of this information. Here, we'll create a "Local" variable where out-of-state offenders are a "0" and those in the state of CA are a "1".

In [15]:
pc["Local"] = np.where(pc["RP State Plate"] == "CA", 1,0)

## 3.4 Within-variable dimensionality reduction
A lot of the categorical variables have dozens of category levels, and many of these levels are sparsely used. When it comes time to one-hot encode these variables, this will cause problems. Here, we'll take a look at a few such variables and make some decisions as to how we can reduce the number of category levels for such variables.

### 3.4.1 Vehicle color
There seems to be some inconsistancy in how the vehicle's color is coded. The most popular vehicle colors can be reasonably deduced from their two-character abbreviations, but numerous obsecure colors appear in the data. We'll arbitrarily select the top eight colors for preservation, and lump all other colors into a ninth "other" category.

In [16]:
pc["Color"].value_counts()

BK    1605994
WT    1566397
GY    1442724
SL     947624
BL     566807
RD     372034
GN     207427
BN     153016
GO      93502
MR      87383
TN      54226
WH      52083
OT      49067
BG      45591
YE      29117
SI      29087
OR      21050
RE      14427
PR      13575
GR       8724
UN       5676
TA       3807
TU       3284
BR       3019
CO       2911
MA       2296
BU       1176
PK       1158
BE        745
PU        330
       ...   
CA          2
BI          1
WA          1
SH          1
WE          1
SN          1
BC          1
VI          1
BA          1
BH          1
AM          1
AP          1
UK          1
W           1
OG          1
O           1
IN          1
UT          1
TW          1
R           1
PP          1
IV          1
RA          1
RS          1
ES          1
KA          1
MP          1
SV          1
TL          1
WR          1
Name: Color, Length: 82, dtype: int64

In [17]:
colors_top8 = pc["Color"].value_counts().index.tolist()[:8]
pc["Color_rev"] = np.where(pc["Color"].isin(colors_top8),pc["Color"],"Other")
pc["Color_rev"].value_counts()

BK       1605994
WT       1566397
GY       1442724
SL        947624
BL        566807
Other     524479
RD        372034
GN        207427
BN        153016
Name: Color_rev, dtype: int64

### 3.4.2 Agency
Upon exploring the value counts of the agency issuing the parking citation, it appears that the vast majority of citations in the data set are written by the top five agencies. As was done with vehicle color, we'll choose the top five issuing agencies and lump the remaining citations under a sixth "Other" agency.

In [18]:
pc["Agency"].value_counts()

54.0    1870363
51.0    1700459
56.0    1626872
53.0    1363992
55.0     670744
1.0       63942
2.0       41477
57.0      17066
4.0       13440
58.0       5615
36.0       5430
11.0       2923
40.0       2834
3.0         910
34.0        267
41.0         69
97.0         25
5.0          18
59.0         14
12.0         10
50.0         10
6.0           7
7.0           2
52.0          2
35.0          1
32.0          1
15.0          1
77.0          1
10.0          1
8.0           1
38.0          1
Name: Agency, dtype: int64

In [19]:
agency_top5 = pc["Agency"].value_counts().index.tolist()[:5]
pc["Agency_rev"] = np.where(pc["Agency"].isin(agency_top5), pc["Agency"], "Other")
pc["Agency_rev"].value_counts()

54.0     1870363
51.0     1700459
56.0     1626872
53.0     1363992
55.0      670744
Other     154072
Name: Agency_rev, dtype: int64

### 3.4.3 Violation Description
The "Violation Description" data are a bit tainted. For example, there is a category labeled "NO STOP/STANDING" and another labeled "NO STOP/STAND". Moreover, because there are some 500+ different violation descriptions, our data set will benefit from a reduction in this dimension. As we've done so far, we'll arbitrarily take the top ten violation descriptions and lump the remaining into an eleventh "Other" category.

In [20]:
pc["Violation Description"].value_counts()

NO PARK/STREET CLEAN              2049213
METER EXP.                        1457978
PREFERENTIAL PARKING               550998
RED ZONE                           537882
DISPLAY OF TABS                    429395
NO PARKING                         334970
DISPLAY OF PLATES                  209778
WHITE ZONE                         177321
PARKED OVER TIME LIMIT             147096
NO STOP/STANDING                   136968
PARKED OVER TIME LIM               111620
STANDNG IN ALLEY                    92572
BLOCKING DRIVEWAY                   89989
YELLOW ZONE                         81856
STOP/STAND PROHIBIT                 81777
NO STOP/STAND                       76389
PARKED ON SIDEWALK                  61085
18 IN. CURB/2 WAY                   57928
FIRE HYDRANT                        57823
EXCEED 72HRS-ST                     52451
OUTSIDE LINES/METER                 52136
NO STOPPING/ANTI-GRIDLOCK ZONE      50355
DOUBLE PARKING                      48584
OFF STR/OVERTIME/MTR              

In [21]:
vd_top10 = pc["Violation Description"].value_counts().index.tolist()[:10]
pc["VD_rev"] = np.where(pc["Violation Description"].isin(vd_top10), pc["Violation Description"], "Other")
pc["VD_rev"].value_counts()

NO PARK/STREET CLEAN      2049213
METER EXP.                1457978
Other                     1354903
PREFERENTIAL PARKING       550998
RED ZONE                   537882
DISPLAY OF TABS            429395
NO PARKING                 334970
DISPLAY OF PLATES          209778
WHITE ZONE                 177321
PARKED OVER TIME LIMIT     147096
NO STOP/STANDING           136968
Name: VD_rev, dtype: int64

## 3.5 Cleaning up
Since we've revised the "Color", "Agency", and "Violation Description" data into new columns, we can drop the original columns.

In [22]:
pc = pc.drop(["RP State Plate", "Body Style", "Color", "Agency", "Violation Description", "Make", "Route"], axis = 1)

In [23]:
print(str(pc.shape))
pc.head()

(7386502, 14)


Fine amount Issue Date Issue time    Latitude  Longitude  Hour  Year  \
2          58.0 2015-12-21      20:55 -118.401151  33.945212    20  2015   
3          68.0 2015-12-26      15:15 -118.401009  33.945212    15  2015   
33         93.0 2015-12-21      10:41 -118.363082  34.185788    10  2015   
42        363.0 2015-12-19      15:50 -118.438092  34.157607    15  2015   
43         68.0 2015-12-26      18:15 -118.311225  34.126932    18  2015   

    Month  Day DayOfWeek  Local Color_rev Agency_rev    VD_rev  
2      12   21    Monday      1        BK      Other     Other  
3      12   26  Saturday      1     Other      Other     Other  
33     12   21    Monday      1        BK      Other  RED ZONE  
42     12   19  Saturday      1     Other      Other     Other  
43     12   26  Saturday      0     Other      Other     Other

#### We are left with 42 features and one target.

# 4. Exploratory Data Analysis

First, let's take a look at the distribution of "Fine amounts"

In [24]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.figure(figsize = (12,8))
ax = sns.kdeplot(pc["Fine amount"].values, shade = True, bw = 5, label = "bw: 5")
ax.set(xlabel='Fine amount ($USD)', ylabel='Kernel Density Estimation')
plt.show()

<Figure size 1200x800 with 1 Axes>

One may look at this distribution and argue that a natural logarithm transformation might be appropriate. However, we'll forego such a transformation under the advisement of this article here:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4120293/

In [25]:
# # Get counts of citations per day of week
# fine_amounts = pc["Fine amount"].value_counts().index.tolist()
# fine_amounts_sorted = sorted(fine_amounts)
# fine_amounts_sorted_str = [str(i) for i in fine_amounts_sorted]
# fine_amounts_str = [str(i) for i in fine_amounts]
# counts_list = pc["Fine amount"].value_counts().tolist()
# fine_counts = dict(zip(fine_amounts_str, counts_list))

# new_fine_counts = dict()
# for key in sorted(fine_counts.keys(), key = fine_amounts_sorted_str.index):
#     new_fine_counts[key] = fine_counts[key]
# new_fine_counts

# # Plot
# output_notebook()
# amounts = list(new_fine_counts.keys())
# counts = list(new_fine_counts.values())
# source = ColumnDataSource(data=dict(amounts=amounts, counts=counts))
# p = figure(x_axis_label = "Day of Week", y_axis_label = "Count", plot_width = 800, plot_height = 600, x_range=amounts, title="Fine amount counts")
# p.vbar(x='amounts', top='counts', width=0.9, legend=None, source=source, line_color = "black")
# p.xaxis.major_label_orientation = math.pi/3
# show(p)

### What does the spatial distribution of parking citations look like?
First, a look at the spatial distribution of parking citations (just on a subset of the data)

In [26]:
google_api = "AIzaSyDQS2K-iAYzzVLSsAOFaK6CDHSzzaOqzbU"

pc_subset = pc.iloc[::250]

# Spatial distribution of citations
output_notebook()
map_options = GMapOptions(lat=34.05, lng=-118.3, map_type="roadmap", zoom=10)
p = gmap(google_api, map_options, title="Los Angeles", x_axis_label = "Latitude", y_axis_label = "Longitude")
p.circle(x=pc_subset["Latitude"].values, y=pc_subset["Longitude"].values, color = "red", alpha = 0.025, size = 10)
show(p)

Loading BokehJS ...

### Are you more or less likely to receive a citaiton on a particular day of the week?

In [27]:
# Get counts of citations per day of week
dayslist = pc["DayOfWeek"].value_counts().index.tolist()
countlist = pc["DayOfWeek"].value_counts().tolist()
daycounts = dict(zip(dayslist,countlist))
order = ['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']
# Sort by day of week
new_daycounts = dict()
for key in sorted(daycounts.keys(), key = order.index):
    new_daycounts[key] = daycounts[key]
new_daycounts

# Plot
output_notebook()
days = list(new_daycounts.keys())
counts = list(new_daycounts.values())
source = ColumnDataSource(data=dict(days=days, counts=counts, color=Spectral7))
p = figure(x_axis_label = "Day of Week", y_axis_label = "Count", plot_width = 800, plot_height = 600, x_range=days, title="Citation Counts: Day of Week")
p.vbar(x='days', top='counts', width=0.9, legend=None, source=source, line_color = "black")
show(p)

Loading BokehJS ...

Here, we can see that much fewer citations are written on the weekends, possibly due to different parking restrictions when compared to weekdays. If we were interested in predicting whether a parking citation was likely to be issued, converting 'Day of Week' to a binary 'Weekday vs. Weekend' variable might help to capture the effect shown above.

### Does the day of the week affect the citation fine amount?

In [28]:
# Get average fine per day of week
fine_daylist = pc.groupby(["DayOfWeek"])["Fine amount"].mean().index.tolist()
fine_finelist = pc.groupby(["DayOfWeek"])["Fine amount"].mean().tolist()
fine_finelist = [round(i,2) for i in fine_finelist]
avgfines = dict(zip(fine_daylist, fine_finelist))
# Sort by day of week
new_dayfines = dict()
for key in sorted(avgfines.keys(), key = order.index):
    new_dayfines[key] = avgfines[key]
new_dayfines

# Plot
output_notebook()
days = list(new_dayfines.keys())
avgfines = list(new_dayfines.values())
source = ColumnDataSource(data=dict(days=days, avgfines=avgfines, color=Spectral7))
p = figure(x_axis_label = "Day of Week", y_axis_label = "Average Citation Fine ($USD)", plot_width = 800, plot_height = 600, x_range=days, title="Avg Citation Fines: Day of Week")
p.vbar(x='days', top='avgfines', width=0.9, legend=None, source=source, line_color = "black")
show(p)

Loading BokehJS ...

While it seems true that you're less likely to receive a citation on weekend days, it doesn't seem like you'll get a break on the fine amount if you do receive a citation! Average fine amounts look consistent between days of the week.

### How does the average fine amount change over time?

In [29]:
# Get average fine per year
years = pc.groupby(["Year"])["Fine amount"].mean().index.tolist()
fines = pc.groupby(["Year"])["Fine amount"].mean().tolist()
fines = [round(i,2) for i in fines]
years_fines = dict(zip(years, fines))

output_notebook()
source = ColumnDataSource(data = dict(years = years, fines = fines))
p = figure(x_axis_label = "Year", y_axis_label = "Avg Fine Amount ($USD)", plot_width = 800, title = "Average Fine Amount per Year", y_range = (50,90))
p.line(x = 'years', y = 'fines', source = source, line_width = 5)
p.circle(x = 'years', y = 'fines', source = source, size = 10, fill_color = "white")
p.circle(x = 'years', y = 'fines', size=10, source = source, fill_color='grey', alpha=0.1, line_color = None, hover_fill_color='firebrick', hover_alpha=0.75)
hover = HoverTool(tooltips = None, mode = 'vline')
p.add_tools(hover)
show(p)

Loading BokehJS ...

In looking at the average fine amount per year over the entire data set, we see a noticeable spike in the average fine amount in 2013, stabilizing thereafter. Perhaps the citation year will indeed help predict the citation amount. Also note that some data for 2019 is also present!

### Is a particular agency prone to writing citations with more hefty fines?
It may be the case that different agencies cover different parts of the city, and thus are prone to writing citations for more expensive offenses. Let's take a look!

In [30]:
# Get average fine per agency
fine_alist = pc.groupby(["Agency_rev"])["Fine amount"].mean().index.tolist()
fine_alist = [str(i) for i in fine_alist]
fine_finelist = pc.groupby(["Agency_rev"])["Fine amount"].mean().tolist()
fine_finelist = [round(i,2) for i in fine_finelist]
avgfines = dict(zip(fine_alist, fine_finelist))


# Plot
output_notebook()
agency = list(avgfines.keys())
avgfines = list(avgfines.values())
source = ColumnDataSource(data=dict(agency=agency, avgfines=avgfines))
p = figure(x_axis_label = "Agency", y_axis_label = "Average Citation Fine ($USD)", plot_width = 800, plot_height = 600, x_range=agency, title="Avg Citation Fines per Agency")
p.vbar(x='agency', top='avgfines', width=0.9, legend=None, source=source, line_color = "black")
show(p)

Loading BokehJS ...

In looking to see whether a particular agency is more inclined to write citations with larger fines, there appears to be some minor fluctuation between agencies. The fact that the "Other" group is so much higher than average may be an artifact from the aggregation steps we took in the feature engineering stage.

### Does vehicle color have anything to do with fine amounts?
At least for moving violations (i.e., not what we're looking at here) there's a popular perception that law enforcement is biased against flashy colors, particularly RED vehicles. Does that hold with regard to parking citations?

In [31]:
# Get average fine per color
fine_clist = pc.groupby(["Color_rev"])["Fine amount"].mean().index.tolist()
fine_clist = [str(i) for i in fine_clist]
fine_finelist = pc.groupby(["Color_rev"])["Fine amount"].mean().tolist()
fine_finelist = [round(i,2) for i in fine_finelist]
avgfines = dict(zip(fine_clist, fine_finelist))


# Plot
output_notebook()
c = list(avgfines.keys())
avgfines = list(avgfines.values())
source = ColumnDataSource(data=dict(c=c, avgfines=avgfines))

mapper = CategoricalColorMapper(
    factors = ["BK","BL","BN","GN","GY","Other","RD","SL","WT"],
    palette = ["black","blue","brown","green","grey","purple","red","silver","white"]
)

p = figure(x_axis_label = "Vehicle Color", y_axis_label = "Average Citation Fine ($USD)", plot_width = 800, plot_height = 600, x_range=c, title="Avg Citation Fines for each Color")
p.vbar(x='c', top='avgfines', width=0.9, legend=None, source=source, color = {'field':'c','transform':mapper}, line_color = "black", alpha = 0.85)
show(p)

Loading BokehJS ...

Here, it looks as though brown vehicles are more likely to receive citations with higher fines, for some odd reason!

### How does the average fine amount fluctuate with the type of violation?
It's reasonable to expect that some parking offenses are more serious than others, warranting a higher fine amount. Let's look at our violation description categories to get a sense of how the average fine amount is associated.

In [32]:
# Get average fine per violation description
fine_vdlist = pc.groupby(["VD_rev"])["Fine amount"].mean().index.tolist()
fine_vdlist = [str(i) for i in fine_vdlist]
fine_finelist = pc.groupby(["VD_rev"])["Fine amount"].mean().tolist()
fine_finelist = [round(i,2) for i in fine_finelist]
avgfines = dict(zip(fine_vdlist, fine_finelist))


# Plot
output_notebook()
vd = list(avgfines.keys())
avgfines = list(avgfines.values())
source = ColumnDataSource(data=dict(vd=vd, avgfines=avgfines))
p = figure(x_axis_label = "Violation Description", y_axis_label = "Average Citation Fine ($USD)", plot_width = 800, plot_height = 600, x_range=vd, title="Avg Citation Fines for each Violation Description")
p.vbar(x='vd', top='avgfines', width=0.9, legend=None, source=source, line_color = "black")
p.xaxis.major_label_orientation = math.pi/3
show(p)

Loading BokehJS ...

It looks like the average fine amount is associated with the type of violation, to some extent.

# 5. Feature Selection

## 5.1 One-HOT encoding
Before we examine our data further, we need to one-HOT encode our categorical variables so that they are numeric. This is necessary for our future models to operate smoothly.

In [33]:
pc = pd.get_dummies(pc, columns = ["DayOfWeek","Color_rev","Agency_rev","VD_rev"])
pc.head()

Fine amount Issue Date Issue time    Latitude  Longitude  Hour  Year  \
2          58.0 2015-12-21      20:55 -118.401151  33.945212    20  2015   
3          68.0 2015-12-26      15:15 -118.401009  33.945212    15  2015   
33         93.0 2015-12-21      10:41 -118.363082  34.185788    10  2015   
42        363.0 2015-12-19      15:50 -118.438092  34.157607    15  2015   
43         68.0 2015-12-26      18:15 -118.311225  34.126932    18  2015   

    Month  Day  Local        ...          VD_rev_DISPLAY OF TABS  \
2      12   21      1        ...                               0   
3      12   26      1        ...                               0   
33     12   21      1        ...                               0   
42     12   19      1        ...                               0   
43     12   26      0        ...                               0   

    VD_rev_METER EXP.  VD_rev_NO PARK/STREET CLEAN  VD_rev_NO PARKING  \
2                   0                            0                  0   
3                   0                            0                  0   
33                  0                            0                  0   
42                  0                            0                  0   
43                  0                            0                  0   

    VD_rev_NO STOP/STANDING  VD_rev_Other  VD_rev_PARKED OVER TIME LIMIT  \
2                         0             1                              0   
3                         0             1                              0   
33                        0             0                              0   
42                        0             1                              0   
43                        0             1                              0   

    VD_rev_PREFERENTIAL PARKING  VD_rev_RED ZONE  VD_rev_WHITE ZONE  
2                             0                0                  0  
3                             0                0                  0  
33                            0                1                  0  
42                            0                0                  0  
43                            0                0                  0  

[5 rows x 43 columns]

## 5.2 Examine correlations with the target variable

In [34]:
corr_data = pc.corr()["Fine amount"].sort_values()
corr_data

VD_rev_DISPLAY OF TABS          -0.355146
VD_rev_DISPLAY OF PLATES        -0.244375
VD_rev_METER EXP.               -0.111031
VD_rev_WHITE ZONE               -0.059991
VD_rev_PARKED OVER TIME LIMIT   -0.054524
Agency_rev_51.0                 -0.030603
Color_rev_BK                    -0.023821
Local                           -0.023786
Agency_rev_54.0                 -0.022807
DayOfWeek_Saturday              -0.022143
VD_rev_PREFERENTIAL PARKING     -0.018534
Hour                            -0.013767
Color_rev_BL                    -0.006196
Color_rev_SL                    -0.005699
Agency_rev_56.0                 -0.005210
Longitude                       -0.002851
Color_rev_GY                    -0.002039
Color_rev_GN                    -0.001972
DayOfWeek_Tuesday               -0.001898
Color_rev_RD                     0.000543
DayOfWeek_Friday                 0.000919
DayOfWeek_Monday                 0.001585
DayOfWeek_Wednesday              0.002634
Latitude                         0

Generally speaking, most of our features are showing rather weak correlations with the fine amount. Noteable acceptions are the different categories of violation description. Let's use a recursive feature elimination algorithm to see if we can eliminate some factors ahead of modeling. But first, we should split our data into training/testing sets...

## 5.3 Train/Test Splitting
Since we have 'Year','Month','Day', and 'Hour' variables to capture the date/time information, we'll omit the 'Issue Date' and 'Issue time' from our analysis for the moment. This is also the appropriate time to use feature scaling.

In [35]:
X, y = pc.iloc[:,3:], pc.iloc[:,0]

scaler = StandardScaler()
X_scale = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_scale, y, test_size = 0.3, random_state = 42)

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


## 5.4 Recursive Feature Elimination

In [36]:
# estimator = LinearRegression(n_jobs = -1)
# selector = RFECV(estimator, cv = 5, scoring = "neg_mean_squared_error")
# selector = selector.fit(X_train,y_train)
# print("Optimal number of features : %d" % selector.n_features_)
# features = list(X.columns[selector.support_].unique())

In [37]:
# print(features)

In [38]:
# pc = pd.concat([pc["Fine amount"], pc[features]], axis = 1)

In [39]:
pc.head()

Fine amount Issue Date Issue time    Latitude  Longitude  Hour  Year  \
2          58.0 2015-12-21      20:55 -118.401151  33.945212    20  2015   
3          68.0 2015-12-26      15:15 -118.401009  33.945212    15  2015   
33         93.0 2015-12-21      10:41 -118.363082  34.185788    10  2015   
42        363.0 2015-12-19      15:50 -118.438092  34.157607    15  2015   
43         68.0 2015-12-26      18:15 -118.311225  34.126932    18  2015   

    Month  Day  Local        ...          VD_rev_DISPLAY OF TABS  \
2      12   21      1        ...                               0   
3      12   26      1        ...                               0   
33     12   21      1        ...                               0   
42     12   19      1        ...                               0   
43     12   26      0        ...                               0   

    VD_rev_METER EXP.  VD_rev_NO PARK/STREET CLEAN  VD_rev_NO PARKING  \
2                   0                            0                  0   
3                   0                            0                  0   
33                  0                            0                  0   
42                  0                            0                  0   
43                  0                            0                  0   

    VD_rev_NO STOP/STANDING  VD_rev_Other  VD_rev_PARKED OVER TIME LIMIT  \
2                         0             1                              0   
3                         0             1                              0   
33                        0             0                              0   
42                        0             1                              0   
43                        0             1                              0   

    VD_rev_PREFERENTIAL PARKING  VD_rev_RED ZONE  VD_rev_WHITE ZONE  
2                             0                0                  0  
3                             0                0                  0  
33                            0                1                  0  
42                            0                0                  0  
43                            0                0                  0  

[5 rows x 43 columns]

# 6. Baseline naive model

In [40]:
from sklearn.metrics import mean_squared_error

def rmse(y_actual, y_predicted):
    mse = mean_squared_error(y_actual, y_predicted)
    rmse = np.sqrt(mse)
    return rmse

In [41]:
baseline_guess = np.median(y_train)

print("Naively estimated Fine Amount: $" + str(baseline_guess))

Naively estimated Fine Amount: $68.0


In [42]:
baseline_guess = [baseline_guess] * len(y_test)
baseline_rmse = rmse(y_test, baseline_guess)

print("Baseline RMSE on test set: $" + str(round(baseline_rmse,2)))

Baseline RMSE on test set: $31.74


# 7. Comparison of Baseline ML models

Let's evaluate a handful of machine learning algorithms with the training data:

    1. Linear Regression
    2. Bayesian Ridge Regression
    3. Lasso Regression
    4. Ridge Regression
    5. Random Forests
    6. Linear Support Vector Regression
    7. XGBoost


In [45]:
baseline_models = pd.DataFrame(columns=[
    "Algorithm",
    "Train RMSE",
    "Train Time"])

baseline_model_scores = dict()

lr = LinearRegression()
bayes = BayesianRidge()
lasso = Lasso()
ridge = Ridge()
rforest = RandomForestRegressor()
linsvr = LinearSVR(random_state = 42)
xgbst = xgb.XGBRegressor(objective = "reg:linear", early_stopping_rounds = 5, tree_method = "approx", seed = 42)

models = [lr, bayes, lasso, ridge, rforest, linsvr, xgbst]
names = ["LR","BayesRidge","Lasso","Ridge","RForest","LinSVR","XGBoost"]

for i in range(len(models)):
    name = names[i]
    t0 = time()
    score = cross_val_score(models[i], X_train, y_train, cv = 5, scoring = "neg_mean_squared_error",n_jobs = -1, verbose=2)
    time_train = round(time() - t0, 2)
    score = np.sqrt(-score)
    baseline_model_scores[name] = score
    rmse = round(np.mean(score),2)
    baseline_models.loc[i] = [name, rmse, time_train]

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:   56.8s remaining:   37.9s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   56.9s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:  1.4min remaining:   57.7s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:  1.4min finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:   37.3s remaining:   24.9s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   38.3s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:   10.6s remaining:    7.1s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   11.3s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parall

In [46]:
baseline_models

Algorithm  Train RMSE  Train Time
0          LR       26.43       56.97
1  BayesRidge       26.43       86.91
2       Lasso       26.63       38.32
3       Ridge       26.43       11.41
4     RForest       18.94      635.17
5      LinSVR       27.56      133.51
6     XGBoost       25.94     1575.60

In [ ]:
baseline_model_scores['RForest']

Based on the findings of these models, it appears that a Random Forest regressor performs best, reducing the RMSE from our baseline of 31.74 to 18.94, followed by XGBoost (25.94). It's worth noting that Linear Support Vector Regression is more sensitive to hyperparameter tuning than other models, so we won't count it out just yet!

# 8. Hyperparameter Tuning - Stage One

Now, we'll take a closer look at three of the seven models tested above: Random Forest Regression, XGBoost, and Linear Support Vector Regression. We will use hyperparameter tuning to optimize each of the three models, and select the best of the three models for further analysis.

In [53]:
import json

# Set up empty lists to capture the model objects and their CV RMSE scores
regressor_objects = []
regressor_scores = []

def ttest(control, treatment):
    from scipy import stats
    
    (t_score, p_value) = stats.ttest_rel(control, treatment)

    if p_value > 0.05/2:  #Two sided
        print('There is no significant difference between the two machine learning pipelines (Accept H0)')
    else:
        print('The two machine learning pipelines are different (reject H0) \n(t_score, p_value) = (%.2f, %.5f)'%(t_score, p_value))
    
    return p_value

In [54]:
# Setup an empty dataframe to capture our hypertuning results
results_RS1 = pd.DataFrame(columns=[
    "ExpID",
    "Train RMSE",
    "p-value",
    "Train Time",
    "Experiment description"])

In [55]:
def ConductRandSearch(X_train, y_train, i=0, prefix=""):
    classifiers = [
        ('LinearSVR', LinearSVR(random_state = 42)),
        ('RandomForest', RandomForestRegressor(random_state = 42)),
        ('XGBoost', xgb.XGBRegressor(objective = "reg:linear", early_stopping_rounds=5, seed = 42, tree_method = "approx"))
    ]

    params_grid = {
        'LinearSVR': {
            'epsilon': (0,0.1,1),
            'tol': (0.0001,0.001,0.01,0.1,1),
            'C': (1,10,100),
            'loss': ('epsilon_insensitive','squared_epsilon_insensitive')
        },
        'RandomForest':  {
            'n_estimators': (10,50,100,250,500),
            'max_features': (1,2,4,8)
        },
        'XGBoost' : {
            'learning rate': (0.1,0.2,0.5,0.8),
            'gamma': (0,10,100,1000), # min loss reduction
            'lambda': (0,1,10,100), # L2 reg
            'alpha': (0,1,10,100), # L1 reg
            'max_depth': (0,2,5,10),
            'subsample': (0.2,0.5,0.8,1) ,
            'colsample_bytree': (0.2,0.5,0.8,1)
        }
    }

    for (name, classifier) in classifiers:
        # Print classifier and parameters
        print(f"****** START {prefix}{name}  *****")
        parameters = params_grid[name]
        print(f"Parameters:")
        for p in sorted(parameters.keys()):
            print(f"\t{p}: {parameters[p]}")

        # Execute the randomized search
        rand_search = RandomizedSearchCV(classifier, parameters, scoring = 'neg_mean_squared_error', n_iter = 5, cv = 5, n_jobs = 4, pre_dispatch = 8, verbose = 2)
        rand_search.fit(X_train, y_train)

        # Best estimator score
        best_train = rand_search.best_score_
        regressor_objects.append(rand_search.best_estimator_)

        # Best estimator fitting time
        t0 = time()
        rand_search.best_estimator_.fit(X_train, y_train)
        time_train = round(time() - t0, 2)

        # Generate 5 training accuracy scores with the best estimator and 5-split CV
        best_train_scores = cross_val_score(rand_search.best_estimator_, X_train, y_train, cv= 5, scoring='neg_mean_squared_error', n_jobs = 4, pre_dispatch = 8, verbose = 2)
        regressor_scores.append([best_train_scores])
        best_train_rmse = np.sqrt(-1*best_train_scores).mean()
        best_train_accuracy = round(best_train_rmse, 2)

        # Conduct t-test with baseline (control) and best estimator (experiment)
        p_value = ttest(baseline_model_scores["RForest"], best_train_scores)

        # Collect the best parameters found by the grid search
        print("Best Parameters:")
        best_parameters = rand_search.best_estimator_.get_params()
        param_dump = []
        for param_name in sorted(parameters.keys()):
            param_dump.append((param_name, best_parameters[param_name]))
            print(f"\t{param_name}: {best_parameters[param_name]}")
        print(f"****** FINISH {prefix}{name}  *****")
        print("")

        # Record the results
        results_RS1.loc[i] = [prefix+name, best_train_rmse, np.round(p_value,3), time_train,json.dumps(param_dump)]

        i += 1

In [ ]:
if __name__ == "__main__":
    ConductRandSearch(X_train, y_train, 0, "Testing: ")

# 9. Hyperparameter Tuning - Stage Two

# 10. Model Evaluation with Test Set

# 11. Model Interpretation with SHAP

# 12. Conclusions